app/models/pending_induction_submission_batch.rb

```diff
@@ -0,0 +1,184 @@
+class PendingInductionSubmissionBatch < ApplicationRecord
+  # Class methods
+  # @return [PendingInductionSubmissionBatch] type "claim"
+  def self.new_claim_for(appropriate_body:, **)
+    new(appropriate_body:, batch_type: 'claim', **)
+  end
+
+  # @return [PendingInductionSubmissionBatch] type "action"
+  def self.new_action_for(appropriate_body:, **)
+    new(appropriate_body:, batch_type: 'action', **)
+  end
+
+  # @return [Class] A class that represents a row in the CSV file
+  # @param columns [Array<Symbol>] Array of CSV column headings
+  def self.build_row_class(columns)
+    Data.define(*columns) do
+      include Enumerable
+
+      # @return [Boolean]
+      def invalid_trn?
+        trn !~ /\A\d{7}\z/
+      end
+
+      # all cells except "error" have content
+      # @return [Boolean]
+      def blank_cell?
+        members[0..-2].any? { |key| public_send(key).blank? }
+      end
+
+      # date cells are formatted as YYYY-MM-DD
+      # @return [Boolean]
+      def invalid_date?
+        members.grep(/dob|_date/).any? do |key|
+          !Date.iso8601(public_send(key))
+        rescue Date::Error
+          true
+        end
+      end
+
+      # @return [Array<String>]
+      def with_errors(error_message)
+        row_values = to_a
+        row_values.delete_at(-1)
+        row_values.push(error_message)
+      end
+
+      # @return [Enumerable]
+      def each(&block)
+        to_a.each(&block)
+      end
+
+      # Guards against Encoding::CompatibilityError i.e. "André"
+      # @return [Array<String>] encoded values for the row
+      def to_a
+        members.map do |key|
+          public_send(key).dup&.force_encoding("UTF-8")
+        end
+      end
+    end
+  end
+
+  CLAIM_CSV_HEADINGS = {
+    trn: 'TRN',
+    dob: 'Date of birth',
+    induction_programme: 'Induction programme',
+    start_date: 'Start date',
+    error: 'Error message',
+  }.freeze
+
+  ACTION_CSV_HEADINGS = {
+    trn: 'TRN',
+    dob: 'Date of birth',
+    end_date: 'End date',
+    number_of_terms: 'Number of terms',
+    objective: 'Objective',
+    error: 'Error message',
+  }.freeze
+
+  ActionRow = build_row_class(ACTION_CSV_HEADINGS.keys)
+  ClaimRow = build_row_class(CLAIM_CSV_HEADINGS.keys)
+
+  # Associations
+  belongs_to :appropriate_body
+  has_many :pending_induction_submissions
+  has_one_attached :csv_file
```

Comment:
I don't really see why we want to hold onto the file here, and active storage adds complexity on the deployment/backup side.

The way it worked in my mind was that the upload request creates pending induction submissions which _represent_ the row.

Or is there a reason to keep the original that I missed?

Reply:
@peteryates the relevant changes in this PR are just the newest commit but I needed it deployed for Joe and Priyah. You'd already commented about blob storage on the original PR and I am replacing active storage.